### DISCUSSION TITLE
Throttling Task.Run

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#62483] [Q&A]
I have a highly concurrent app that transfers data between systems.
I would like to implement throttling in it so that it does not bombard the source/dest systems.
I think the best way to do this is with a custom thread pool that is limited to X number of concurrent threads.

How do I create a custom thread pool and ensure that all Task.Runs from inside that thread pool stay on that thread pool?

**Comment by gfoidl** [comment]
See e.g. [Limiting concurrency for faster and more responsive apps](https://devblogs.microsoft.com/premier-developer/limiting-concurrency-for-faster-and-more-responsive-apps/) which uses the `ConcurrentExclusiveSchedulerPair` as _TaskScheduler_.

Or you could limit the work with a `Semaphore` (or the *Slim variant), depending on your actual use-case.
Or use a Producer/Consumer like Channels, TPL Dataflow, etc. to limit the amount of work that can be done "at once".
Or ... there are many possibilites, hard to say without knowing your system.
